src/compiler/checker.ts

```diff
@@ -9072,8 +9072,9 @@ namespace ts {
                     // If the parent is a tuple type, the rest element has a tuple type of the
                     // remaining tuple element types. Otherwise, the rest element has an array type with same
                     // element type as the parent type.
-                    type = everyType(parentType, isTupleType) ?
-                        mapType(parentType, t => sliceTupleType(t as TupleTypeReference, index)) :
+                    const resolvedParentType = getBaseConstraintOrType(parentType);

```

Comment:
This might already be an improvement but it feels that there is also a bigger improvement to be had here. This should somehow use the flow type - although it might not be simple as `getFlowTypeOfDestructuring` and related functions don't support dotted declarations~.

Maybe worth reporting an issue about it (unless we find an existing one).

Reply:
Seems reasonable. I'll let you file that issue 😄

Reply:
actually - maybe this isn't an issue since this particular union is not a discriminated one, I'll try to recheck that later